Sources/_MatchingEngine/Regex/AST/AST.swift

```diff
@@ -207,19 +210,43 @@ extension AST {
     }
     public var kind: Kind
 
+    /// An additional specifier supported by Oniguruma that specifies what
+    /// recursion level the group being referenced belongs to.
+    public var recursionLevel: Located<Int>?
+
     /// The location of the inner numeric or textual reference, e.g the location
-    /// of '-2' in '\g{-2}'.
+    /// of '-2' in '\g{-2}'. Note this includes the recursion level for e.g
+    /// '\k<a+2>'.
     public var innerLoc: SourceLocation
 
-    public init(_ kind: Kind, innerLoc: SourceLocation) {
+    public init(_ kind: Kind, recursionLevel: Located<Int>? = nil,
+                innerLoc: SourceLocation) {
       self.kind = kind
+      self.recursionLevel = recursionLevel
       self.innerLoc = innerLoc
     }
 
     /// Whether this is a reference that recurses the whole pattern, rather than
     /// a group.
     public var recursesWholePattern: Bool { kind == .recurseWholePattern }
   }
+
+  /// An AST node containing global matching options along with a child that
+  /// uses those options.
+  /// TODO: Should this be subsumed into a top-level AST type?
```

Comment:
I think it should. The AST is recursive, but this cannot be a member. So I think it might be time for `struct AST` and `enum AST.Node`.

Reply:
Can you try prototyping that direction? Feel free to split the global options out of this PR if it helps land stuff.

Reply:
Yeah I can work on implementing that. I have committed a bunch of stuff on top of this though, so would have to work through a rebase to split off the global options commit. Would you mind if I landed it in its current state and moved it onto `AST` in a follow-up?

Reply:
@milseman To be clear, I mean that I have local commits based on top of this that I'd need to rebase. I'm happy to do that, but it'd be more convenient for me to land this as-is.